src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -463,5 +463,43 @@ internal static bool IsVoidOrUnit(Type type)
 
             return type == typeof(void) || type.FullName == "Microsoft.FSharp.Core.Unit";
         }
+
+        /// <summary>
+        /// Returns the property information for a given named indexer.
+        /// </summary>
+        /// <param name="type">Type to reflect on for indexer.</param>
+        /// <param name="indexerTypes">List of indexer types that matches the indexer type order.</param>
+        /// <returns>Invokable indexer method</returns>
+        public static MethodInfo? GetIndexer(Type type, Type[] indexerTypes)
+        {
+            const BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance;
+            
+            var indexer = GetDefaultIndexer(type, indexerTypes, bindingFlags) ?? GetNamedIndexer(type, indexerTypes, bindingFlags);
+            
+            return indexer?.GetGetMethod(true);
+        }
+
+        private static PropertyInfo? GetDefaultIndexer(Type type, Type[] indexerTypes, BindingFlags bindingFlags)
+        {
+            const string defaultFrameworkIndexerName = "Item";
+            return type.GetProperty(defaultFrameworkIndexerName, bindingFlags, null, null, indexerTypes, null);
+        }
+
+        private static PropertyInfo? GetNamedIndexer(Type type, Type[] indexerTypes, BindingFlags bindingFlags)
+        {
+            var props = type.GetProperties(bindingFlags).ToArray();
+            return props.FirstOrDefault(p => ParametersEqual(p.GetIndexParameters(), indexerTypes));
```

Comment:
This would not be correct. This would find `non-default indexers`. The correct way to find the *default indexer* that has a *non-default name* would be to check for `DefaultMemberAttribute` on `type`.

If `DefaultMemberAttribute` is present, we should not search for an indexer with a metadata name of `"Item"`. Instead, we should search for the member name specified by the DefaultMemberAttribute.

Reply:
There's a difference between each of these things:

- A default indexer with the default name ("Item"). Because this is a default indexer, you can call it in C# or VB without using the property name. No `DefaultMemberAttribute` is present.
- A default indexer with a non-default name (specified in C# using `IndexerNameAttribute` on the default indexer). Because this is a default indexer, you can call it in C# or VB without using the property name. A `DefaultMemberAttribute` is present.
- A so-called *named indexer*. Because this is not a default indexer, you *can't* call it in C# or VB without using its metadata name. No `DefaultMemberAttribute` is present.

Reply:
I can give code examples of the three things if it helps.

Reply:
Therefore we should have a method called `Reflect.GetDefaultIndexer` and no method referring to named indexers. It should do one search for the default indexer using the name it got from DefaultMemberAttribute, or using the name `Item` only if there is no DefaultMemberAttribute on the declaring type.

Reply:
Ok, so I misunderstood the _ignoring_ of 3 ... not that I realised 3 was a real thing. 

Could you please clarify your last statement -
> It should do one search for the default indexer using the name it got from DefaultMemberAttribute

Is there any preference on searching default `Item` before `DefaultMemberAttribute` or the other way round?

It is just a point of wondering if there is a critical difference in the direction of the search. I would expect (in terms of user cases) having a default overload using `Item` would be far more common, thus search for it first?

Reply:
Specifically, there is one search, one name, that can possibly be a default indexer for a given class. An indexer named `Item` is not a default indexer if `DefaultMemberAttribute` specifies another name. Such an indexer falls into category 3, an indexer whose name must be given when you call it. Not having to use the name when you call it is what makes a thing a default indexer rather than a named indexer.

So it would be wrong to use `Item` if DefaultMemberAttribute specifies another name.

Reply:
Pushed up the implementation to go grab the `DefaultMemberAttribute` method name off the type provided. This now attempts to get the attribute, if not available, defaults back to `Item`. 

Using that name - a search is done using `type.GetProperty(name, ...)`

Reply:
Perfect!